### PR TITLE
[DLLM] Add fused Triton post-process kernel with on-device fallback

### PR DESCRIPTION
**Pull request by YJMSTR**
## Motivation

The dLLM (Diffusion LLM) low-confidence decoding loop performs multiple PyTorch operations per iteration — softmax, argmax, gather, where, threshold comparison, and conditional token update — each launching a separate CUDA kernel. Additionally, per-block `.item()` calls introduce GPU-CPU synchronization overhead. This PR fuses all post-processing into a single Triton kernel and eliminates per-block sync.

## Modifications

- **New `python/sglang/srt/dllm/kernels/` package** with fused Triton kernel (`post_process.py`):
  - Online softmax algorithm: single-pass argmax + probability computation
  - On-device fallback: atomic transfer counter + lightweight fallback kernel (no GPU-CPU sync when no token exceeds threshold)
  - Autotuning across vocab sizes (`BLOCK_V`: 1024–8192, `num_stages`: 2–4)
  - Early exit for non-masked positions
  - PyTorch reference implementation preserved for correctness verification

- **Modified `low_confidence.py`**:
  - Uses Triton kernel by default (`use_triton_post_process=True`)
  - Removed per-block `.item()` sync; single global mask check per iteration
  - Original PyTorch path preserved as fallback

## Accuracy Tests

GSM8K 5-shot (200 questions, LLaDA2.0-mini, H20 GPU, max_running_requests=4):

| Branch | Accuracy |
|---|---|
| main | 0.920 |
| Triton kernel | 0.925 |

Unit tests (`test/registered/dllm/test_dllm_triton_kernel.py`):
- Kernel-level: all-masked, no-masked, mixed, fallback scenarios
- Full-loop: single-batch and multi-batch correctness vs PyTorch reference

### Kernel-Level (RTX 5070 Laptop GPU, block_size=32, vocab_size=128000)

| Implementation | Latency | Speedup |
|---|---|---|
| PyTorch Reference | ~1.20 ms | 1.0x |
| Triton Fused | ~0.13 ms | **9.3x** |

Measured with `benchmark/dllm/bench_triton_post_process.py`.

### End-to-End (LLaDA2.0-mini, H20 GPU, max_running_requests=4, cuda_graph_bs=1,2,3,4)

Measured with `test/registered/dllm/test_llada2_mini.py`.

Single request (bs=1, `test_bs_1_speed`):

| Branch | Speed | Speedup |
|---|---|---|
| main | 303.01 token/s | 1.0x |
| Triton kernel | 317.01 token/s | **+4.6%** |

GSM8K (200 questions, 128 parallel, `test_gsm8k`):

| Branch | Throughput | Speedup |
|---|---|---|
| main | 298.74 token/s | 1.0x |
| Triton kernel | 321.24 token/s | **+7.5%** |

The Triton path eliminates per-block `.item()` GPU-CPU synchronization, which yields larger gains under concurrent batching.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review Process

1. Ping Merge Oncalls to start the PR flow. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - `/tag-run-ci-label`, `/rerun-failed-ci`, `/tag-and-rerun-ci`
4. After green CI and required approvals, ask Merge Oncalls to merge.

---
### CI States

Latest PR Test (Base): :x: [Run #27603118322](https://github.com/sgl-project/sglang/actions/runs/27603118322)
Latest PR Test (Extra): :x: [Run #27603118450](https://github.com/sgl-project/sglang/actions/runs/27603118450)

### CODE REVIEW

**Comment by YJMSTR** [comment]
@ClawSeven @btw616 Could you help review this PR?

File: python/sglang/srt/dllm/algorithm/low_confidence.py [hunk 1]
```diff
@@ -19,6 +20,9 @@ def __init__(
     ):
         super().__init__(config)
         self.threshold = config.algorithm_config.get("threshold", 0.95)
+        self.use_triton_post_process = config.algorithm_config.get(
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/algorithm/low_confidence.py]: I suggest we rename the parameter from "use_triton_post_process" to "disable_fused_triton_algorithm", with a default value of false.

The reason is that this flag is intended for researchers who wish to use custom decoding algorithms instead of the fused kernel. However, in most typical scenarios, we should keep the fused kernel enabled by default.

Additionally, the fused Triton kernel should only be used when both of the following conditions are met:
`disable_fused_triton_algorithm is false`, and
Triton is supported `support_triton(get_global_server_args().attention_backend) is true`.

File: python/sglang/srt/dllm/algorithm/low_confidence.py [hunk 2]
```diff
@@ -49,45 +53,56 @@ def run(
...
-                confidence = torch.where(block_mask_index, p, -np.inf)
-
-                transfer_index = confidence > self.threshold
-
-                if transfer_index.sum().item() == 0:
-                    _, select_index = torch.topk(confidence, k=1)
-                    transfer_index[select_index] = True
-
-                block_input_ids[transfer_index] = x[transfer_index]
+                if self.use_triton_post_process:
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/algorithm/low_confidence.py]: rename use_triton_post_process

File: python/sglang/srt/dllm/algorithm/low_confidence.py [hunk 3]
```diff
@@ -49,45 +53,56 @@ def run(
...
+                    dllm_post_process_fused(
+                        curr_logits,
+                        block_input_ids,
+                        self.mask_id,
+                        self.threshold,
+                    )
+                else:
+                    # PyTorch fallback
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/algorithm/low_confidence.py]: The annotation should be "Customized PyTorch Algorithm Implementation"

File: python/sglang/srt/dllm/kernels/post_process.py [hunk 4]
```diff
@@ -0,0 +1,239 @@
...
+        transfer_index,
+        num_transfers_dev,
+        block_size,
+    )
+
+
+dllm_post_process = dllm_post_process_fused
+
+
+def dllm_post_process_pytorch(
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/kernels/post_process.py]: Could you please remove this function, as it is almost identical to the main Python implementation path?

File: test/registered/dllm/test_dllm_triton_kernel.py [hunk 5]
```diff
@@ -0,0 +1,284 @@
+from sglang.test.ci.ci_register import register_cuda_ci
+
+register_cuda_ci(est_time=60, suite="stage-b-test-large-1-gpu")
```
Comment (ClawSeven) [inline test/registered/dllm/test_dllm_triton_kernel.py]: Maybe this PR is not suitable for stage-b-test-large-1-gpu ?

File: python/sglang/srt/dllm/kernels/__init__.py [hunk 6]
```diff
@@ -0,0 +1,13 @@
+"""dLLM post-processing kernels."""
+
+from sglang.srt.dllm.kernels.post_process import (
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/kernels/__init__.py]: Remove to `fused_triton_algorithm`

File: python/sglang/srt/dllm/kernels/low_confidence_utils.py [hunk 7]
```diff
@@ -0,0 +1,239 @@
...
+
+from typing import Tuple
+
+import torch
+import triton
+import triton.language as tl
+
+
+@triton.jit
+def _dllm_post_process_kernel(
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/kernels/low_confidence_utils.py:49]: Since this function is only used for low-confidence cases, I suggest we rename both the file and the function to something more specific, like low_confidence_utils.py and calculate_low_confidence_score.

File: test/registered/dllm/test_dllm_triton_kernel.py [hunk 8]
```diff
@@ -0,0 +1,284 @@
...
+from dataclasses import dataclass
+from typing import Optional
+
+import numpy as np
+import torch
+
+from sglang.srt.dllm.kernels.post_process import (
+    dllm_post_process_fused,
+    dllm_post_process_pytorch,
+)
```
Comment (ClawSeven) [inline test/registered/dllm/test_dllm_triton_kernel.py]: I recommend moving this unit test to the nightly test suite or just delete it. Considered to enable this option by default, the CI run of llada2-test will then automatically validate the Triton kernel each time.

File: test/registered/dllm/test_dllm_triton_kernel.py [hunk 9]
```diff
@@ -0,0 +1,284 @@
...
+from sglang.test.test_utils import CustomTestCase
+
+BLOCK_SIZE = 32
```
Comment (ClawSeven) [inline test/registered/dllm/test_dllm_triton_kernel.py]: Remove those hard code values

File: test/registered/dllm/test_dllm_triton_kernel.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+            self.vocab_size,
+            dtype=torch.float16,
+            device=device,
+            generator=gen,
+        )
+        logits_output = MockLogitsProcessorOutput(full_logits=logits)
+        return MockModelRunnerOutput(logits_output=logits_output, can_run_graph=True)
+
+
+def _run_loop_pytorch(
```
Comment (ClawSeven) [inline test/registered/dllm/test_dllm_triton_kernel.py]: Please delete this branch

File: python/sglang/srt/dllm/kernels/low_confidence_utils.py [hunk 11]
```diff
@@ -0,0 +1,239 @@
...
+            best_conf = c
+            best_idx = i
+
+    argmax_token = tl.load(argmax_out_ptr + best_idx)
+    tl.store(input_ids_ptr + best_idx, argmax_token)
+    tl.store(transfer_out_ptr + best_idx, tl.cast(1, tl.int8))
+    tl.store(num_transfers_ptr, 1)
+
+
+_dllm_post_process_kernel_autotuned = triton.autotune(
```
Comment (ClawSeven) [inline python/sglang/srt/dllm/kernels/low_confidence_utils.py:144]: Could you remove those hard code values?

**Review by ClawSeven — CHANGES_REQUESTED** [review]

**Comment by ClawSeven** [comment]
I think this PR contains a lot of unnecessary code. I'll re-review it once you've addressed the previous comments.

Also, please using less AI-assisted coding, as it tends to introduce redundant code.

File: python/sglang/srt/dllm/kernels/low_confidence_utils.py (hunk 11 above)
Comment (YJMSTR) [reply python/sglang/srt/dllm/kernels/low_confidence_utils.py:144]: Hi, I want to confirm the expected direction so I can align exactly with your preference.
Do you mean one of the following:
1. Keep autotune, but make the config space platform-aware (e.g., different num_stages/num_warps for HIP vs CUDA / different SM), instead of one fixed candidate set for all devices.
2. Keep the current candidate set, but refactor all magic numbers into named constants/helper functions for readability and maintainability.
3. Move tuning values to an external config/override mechanism (similar to fused MoE config style), and keep only a minimal fallback in code.

File: test/registered/dllm/test_dllm_triton_kernel.py (hunk 8 above)
Comment (YJMSTR) [reply test/registered/dllm/test_dllm_triton_kernel.py]: deleted

File: python/sglang/srt/dllm/kernels/low_confidence_utils.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+    confidence_ptr,
+    argmax_out_ptr,
+    transfer_out_ptr,
+    num_transfers_ptr,
+    block_size,
+):
+    num_transfers = tl.load(num_transfers_ptr)
+    if num_transfers > 0:
+        return
+
```
Comment (fdz-1999) [inline python/sglang/srt/dllm/kernels/low_confidence_utils.py:129]: Performance Issue : This kernel is launched with grid=(1,), but Triton defaults to num_warps=4 (128 threads). All threads execute identical code redundantly, resulting in 128x wasted computation and redundant memory writes.
Suggestion: Specify num_warps=1 in the kernel launch to minimize redundant execution.

File: python/sglang/srt/dllm/kernels/low_confidence_utils.py (hunk 11 above)
Comment (fdz-1999) [reply python/sglang/srt/dllm/kernels/low_confidence_utils.py:144]: Direction 2 has been applied — the autotune values are now extracted into named 
constants (_AUTOTUNE_BLOCK_V_OPTIONS, _AUTOTUNE_NUM_WARPS_OPTIONS, etc.) with 
clear semantic meaning.

These values are not arbitrary hardcodes — they represent the standard Triton 
tuning search space for this kernel's compute pattern (vocabulary-dimension 
softmax + threshold comparison). The current 4×2×~3 ≈ 24 configs should cover 
the performance curve inflection points; if we find gaps in practice (e.g., a 
specific vocab_size range or GPU architecture where none of the current configs 
perform well), additional candidates can be added incrementally. The fallback 
logic provides production-safe defaults that skip autotune entirely when needed.

Direction 1 (platform-aware config pruning) and Direction 3 (external config 
mechanism) are reasonable future improvements, but the current setup should 
serve the initial integration well. We can revisit if cold-start or 
device-specific tuning becomes a pain point.

LGTM, resolving.

File: python/sglang/srt/dllm/kernels/low_confidence_utils.py (hunk 12 above)
Comment (fdz-1999) [reply python/sglang/srt/dllm/kernels/low_confidence_utils.py:129]: no blocking

**Comment by ClawSeven** [comment]
@YJMSTR, apologies for the delayed response. Our team has been focused on internal deadlines, leading to a hold on this PR. We appreciate your contribution and patience. The feedback has been provided, and we are prioritizing the merge of this PR.

**Comment by YJMSTR** [comment]
@ClawSeven @fdz-1999 Thanks for the previous reviews and your patience. I have rebased onto the latest upstream/main and force-pushed. All review comments have been addressed.

**Review by ClawSeven — APPROVED** [review]
